File: solutions/ocaml/bob/1/bob.ml
open Base

let response_for text =
  let is_question text = text |> String.strip |> String.is_suffix ~suffix:"?" in
  let is_shout text =
    String.exists text ~f:Char.is_uppercase
    && not (String.exists text ~f:Char.is_lowercase)
  in
  match (is_question text, is_shout text) with
  | true, false -> "Sure."
  | false, true -> "Whoa, chill out!"
  | true, true -> "Calm down, I know what I'm doing!"
  | _, _ when String.is_empty (String.strip text) -> "Fine. Be that way!"
  | _, _ -> "Whatever."

File: solutions/clojure/bob/1/src/bob.clj
(ns bob
  (:require [clojure.string :as s]))

(defn question?
  [s]
  (boolean (re-find #"\?$" s))) ; has "?" at the end

(defn yell?
  [s]
  (and (re-find #"[A-Z]+" s)    ; has any letter
       (= s (s/upper-case s)))) ; all letters are uppercase

(defn yell-question?
  [s]
  (and (question? s)
       (yell? s)))

(defn silence?
  [s]
  (boolean (re-find #"^\s*$" s))) ; whole string is zero or more whitespaces

(defn response-for [s]
  (let [s (clojure.string/trim s)]
      (cond
        (silence? s)       "Fine. Be that way!"
        (yell-question? s) "Calm down, I know what I'm doing!"
        (question? s)      "Sure."
        (yell? s)          "Whoa, chill out!"
        :else              "Whatever.")))
